README.md
# MSFS-PBR
Microsoft Flight Simulator 2020 - Push Back Recorder

Push Back Recorder is an MSFS2020 Push Back utility add-on which allows you to record and playback your push backs


## What can i do with Push Back Recorder?
- Auto-Push Back: Select a recorded PBR file and playback your push backs
- Planned Push Back: Select from Short/Medium/Long, Left/Right, 75°/90°/180° Pre-Planned Push back options
- Manual Push Back: Use your keyboard keys to perform a manual push back
- Record Push Back: Record your push back using your keyboard keys and save your PBR file so it can be played back with the Auto-Push Back feature later on
- Toggle / Push Back: Toggles the push back tug
- Toggle / Jetway: Toggles the jetway connection (If available at your spot)
- Toggle / Brakes: Toggles the Parking Brakes


## Installation 
- Run pbrsetupxxx.exe and complete the install wizard
- Please keep the default install path to avoid issues with the log generator 


## Requirements (If you plan to use the source code)
- [Python 3.8.7](https://www.python.org/downloads/release/python-387/)
- [Python-SimConnect](https://github.com/odwdinc/Python-SimConnect)
- [keyboard](https://github.com/boppreh/keyboard)
- [PyInstaller](https://github.com/pyinstaller/pyinstaller)
- [Inno Setup](https://jrsoftware.org/isinfo.php)


## Known issues 
- PBR won't work if SimRate is above x1.0
- Log error: Could not create pbr.log (Use the default install path to avoid this issue)


## Special thanks to [OverKill](https://www.youtube.com/channel/UC3VY4ObSBiPjHJ9Z5NJ1sAw)
for his collaboration on those amazing [audio files](https://github.com/RushScript/MSFS-PBR/tree/main/Assets/Audio)!


CHANGES
- CHANGES IN THIS VERSION:


[1.0.2]

-- [FIX] Planned push back not responding and keeps going straight (This might have been sorted, please report if the problem persists)

-- [FIX] Fine-tuned push back templates values to avoid static movement at the end of push back

-- [FIX] The app wont run in the background anymore after closing the GUI

-- [ADD] Added Parking Brakes toggle button to the GUI

-- [ADD] You can now use your PBR files in different spots (An information message will be displayed letting you know that the PBR file belongs to a different spot)




[1.0.1]

-- [FIX] Fixed audio comms playing prior Tug inbound

-- [FIX] Fixed audio will stil playing on the background even if the app is closed


[1.0]

-- [FIX] Fixed PBR showing Tip message in MSFS2020 "Recording push back..." while using Manual Push back and Auto-Push back feature

-- [FIX] Record Push Back being not able to record 

-- [FIX] Auto-Push back GUI not working after recording (There is no need to reset the PBR App anymore to use any other feature)

-- [FIX] Push back tug releases the plane without slowing down first (It will now slowdown the plane on release)

-- [FIX] Changed installer compression to zip to avoid AV false positives

-- [FIX] Installer shortcuts fixed (Users can now pin the icon to the taskbar)

-- [FIX] PBR now remembers the last accessed path

-- [IMPROVEMENT] Improved GUI performance

-- [IMPROVEMENT] Improved GUI overall design 

-- [IMPROVEMENT] Push Back Recorder and Play Back fully re-worked

-- [IMPROVEMENT] Play Back accuracy improved

-- [IMPROVEMENT] Reduced CPU usage

-- [ADD] PBR is now compiled as an executable file and doesen't require Python 3.8.7 to be installed on the user system 

-- [ADD] Added new app icon

-- [ADD] Added GUI Menu bar

-- [ADD] Tooltips can now be disabled (Check for settings on the UI menu bar)

-- [ADD] Play Back feature can now read Push Back Templates

-- [ADD] NEW FEATURE: Planned Push Back

-- [ADD] Ground to Flight deck comms audio added during push back phases (Thanks to OverKill for his collaboration on those amazing audio files!)

-- [ADD] Tug will now connect and wait for user to release parking brakes before commencing push back

-- [ADD] Tug will wait for user to set parking brakes before disconnecting 


gui.py
from tkinter import * 

# Button 1 function
def btnClickFunction():
    btn1 = Button(root, text='Auto-Push Back (NO DATA FOUND)', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction).place(x=5, y=12)
    print('Auto Push Back')


# Button 2 function
def btnClickFunction2():
    btn2 = Button(root, text='Starting Push Back...', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction2).place(x=5, y=52)
    print('Start Push Back')


# Button 3 function
def btnClickFunction3():
    btn3 = Button(root, text='Recording Push Back...', state=DISABLED, bg='red', fg="black", font=('arial', 10, 'normal'), command=btnClickFunction3).place(x=5, y=92)
    print('Record Push Back')

root = Tk()

# This is the section of code which creates the main window
root.geometry('395x390')
root.configure(background='#6B6B6B')
root.title('Push Back Recorder')
root.resizable(width=False, height=False)
root.attributes('-topmost', True)
root.update()


# GUI Buttons
btn1 = Button(root, text='Auto-Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction).place(x=5, y=12)
btn2 = Button(root, text='Start Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction2).place(x=5, y=52)
btn3 = Button(root, text='Record Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction3).place(x=5, y=92)


# GUI Canvas images
HowToImg= Canvas(root, height=190, width=380, bd=0, highlightthickness=0, relief='ridge')
picture_file = PhotoImage(file = 'howto.gif')
HowToImg.create_image(380, 0, anchor=NE, image=picture_file)
HowToImg.place(x=5, y=152)

Logo= Canvas(root, height=100, width=100, bd=0, highlightthickness=0, relief='ridge')
picture_fileL = PhotoImage(file = 'logo.gif')  
Logo.create_image(100, 0, anchor=NE, image=picture_fileL)
Logo.place(x=285, y=0)


# GUI Labels
lbl1 = Label(root, text='* Remember to keep this window focused while using manual/record push back', bg='#6B6B6B', fg="#ffc000", font=('arial', 7, 'normal')).place(x=5, y=362)
lbl2 = Label(root, text='SimConnect: Not linked', bg='#6B6B6B', fg="#ffc000", font=('arial', 9, 'normal')).place(x=250, y=83)

root.mainloop()




Main.py
from SimConnect import *
import logging
import math
import time
import keyboard


#Logging configuration
#logging.basicConfig(filename='mhaf.log', filemode='w', level=logging.DEBUG)
logging.basicConfig(level=logging.DEBUG)
LOGGER = logging.getLogger(__name__)
LOGGER.info("START")



def pb():
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    keyboard.add_hotkey("left", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-27)%360*11930464)))
    keyboard.add_hotkey("left", lambda:heading('left'))
    keyboard.add_hotkey("right", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+27)%360*11930464)))
    keyboard.add_hotkey("right", lambda:heading('right'))
    keyboard.add_hotkey("up", lambda:pbst())
    keyboard.add_hotkey("down", lambda:tgl())
    keyboard.add_hotkey("down", lambda:pbkstate())
    return

def pbrec():
    global recdata
    recdata = []
    recdata.append(str(aq.get("PLANE_LATITUDE"))+"\n")
    recdata.append(str(aq.get("PLANE_LONGITUDE"))+"\n")
    recdata.append(str(aq.get("PLANE_HEADING_DEGREES_TRUE"))+"\n")
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    pbr = open(recdata[0].replace('\n', '').replace('.', '')+recdata[1].replace('\n', '').replace('.', '')+'.pbr', 'w')
    keyboard.add_hotkey("left", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-27)%360*11930464)))
    keyboard.add_hotkey("left", lambda:heading('left'))
    keyboard.add_hotkey("right", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+27)%360*11930464)))
    keyboard.add_hotkey("right", lambda:heading('right'))
    keyboard.add_hotkey("up", lambda:pbst())
    keyboard.add_hotkey("down", lambda:tgl())
    keyboard.add_hotkey("down", lambda:recstate())
    while rec > 0:
        if aq.get("GROUND_VELOCITY") > 0.1:
            time.sleep(rft)
            recdata.append(str(int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE"))%360*11930464))+"\n")
    pbr.writelines(recdata)
    pbr.close()
    sm.sendText("Push back recorded and saved to file")
    return

def recstate():
    global rec
    rec = rec - 1
    if (rec == 1):
        sm.sendText("Recording Push back...")
    return rec

def pbplay(fpath):
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    pbrd = open(fpath, 'r') 
    Lines = pbrd.readlines()
    aq.set("PLANE_LATITUDE", float(Lines[0]))
    aq.set("PLANE_LONGITUDE", float(Lines[1]))
    aq.set("PLANE_HEADING_DEGREES_TRUE", float(Lines[2]))
    time.sleep(2)
    sm.sendText("Release parking brakes")
    tgl()
    count = 0
    contact = 1
    while contact > 0:
        if aq.get("GROUND_VELOCITY") > 0.1:
            contact = 0
    for line in Lines:
        count += 1
        if count > 3:
            time.sleep(rft)
            tug(int(line.strip()))
    time.sleep(erft)
    sm.sendText("Set parking brakes")
    tgl()
    

def pbst():
    tug = ae.find("KEY_TUG_HEADING")
    if (hdg == "left"):
        tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-3)%360*11930464))
    elif (hdg == "right"):
        tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+3)%360*11930464))
    return

def heading(trn):
    global hdg
    hdg = trn
    return hdg


def pbkstate():
    global pbstp
    if (pbstp == 0):
        sm.sendText("Release parking brakes")
        pbstp = 1
    elif (pbstp == 1):
        sm.sendText("Set parking brakes")
        pbstp = 0

def simconnectLink() :
    global sm
    #Local var used to loop on SimConnect link attempt
    smconnected = 0
    logging.info("SimConnect:Linking to MSFS2020...")
    while smconnected < 1:
        try:
            sm = SimConnect()
            smconnected = 1
            logging.info("SimConnect:Linked")
            return (sm)
        except:
            logging.info("Retrying in 10sec")
            time.sleep(10)
            continue


sm = simconnectLink()


aq = AircraftRequests(sm) #Aircraft Requests Variable
ae = AircraftEvents(sm) # Aircraft Event Variables

tugtgl = ae.find("TOGGLE_PUSHBACK")
spd = ae.find("KEY_TUG_SPEED")
hdg = ""
pbstp = 0
rec = 2
rft = 0.300
erft = 15


####    print (aq.get("PLANE_LATITUDE"))
####    print (aq.get("PLANE_LONGITUDE"))

##    
##

pbplay("3615306230296912-5347122154676827.pbr")


Gui-mainWIP.py
from tkinter import *
from tkinter import dialog
from SimConnect import *
import os
import logging
import math
import time
import threading
import keyboard

# Logging configuration
# logging.basicConfig(filename='pbr.log', filemode='w', level=logging.DEBUG)
logging.basicConfig(level=logging.DEBUG)
LOGGER = logging.getLogger(__name__)
LOGGER.info("START")


## GUI functions
# Custom dialog function
def customDialog(title, text, strings=('YES', 'NO'), bitmap='question', default=1):
    d = dialog.Dialog(title=title, text=text, bitmap=bitmap, default=default, strings=strings)
    print (strings[d.num])
    return strings[d.num]

# Button 1 function
def btnClickFunction():
    btn2.destroy()
    btn3.destroy()
    HowToImg.destroy()
    #customDialog("Push Back Recorder", "Please select Tail direction for your Push Back", strings=('Left', 'Straight', 'Right'))
    btn1ab.place(x=35, y=52)
    btn2ab.place(x=75, y=52)
    btn3ab.place(x=140, y=52)
    HowToImgTail.place(x=5, y=152)
    logging.info("Auto-Push back PBR file loaded: ")
    btn1["state"] =  DISABLED
    btn1["bg"] = "#ffc000"
    btn1["fg"] = "#6B6B6B"
    btn1["text"] = "Select Auto-Push Back tail direction"


# Button 2 function
def btnClickFunction2():
    if aq.get("GROUND_VELOCITY") > 0.1:
        tugtgl()
        btn2 = Button(root, text='Start Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction2).place(x=5, y=52)
    elif aq.get("GROUND_VELOCITY") < 0.1:    
        btn1 = tkinter.Button(root, text='Auto-Push Back', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction).place(x=5, y=12)
        btn3 = Button(root, text='Record Push Back', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction3).place(x=5, y=92)
        btn2 = Button(root, text='Starting Push Back... (Stop Push Back)', bg='#ffc000', fg="black", font=('arial', 10, 'normal'), command=btnClickFunction2).place(x=5, y=52)
        print('Start Push Back')
        pb()


# Button 3 function
def btnClickFunction3():
    global fpath
    tdct = customDialog("Push Back Recorder", "Please select Tail direction of your Push Back Record", strings=('Left', 'Straight', 'Right'))
    tdct = tdct[0]
    fpath = os.path.dirname(os.path.realpath(__file__))+"\\PBR Data\\longitudelatittude"+tdct+".pbr"
    btn1["state"] =  DISABLED
    btn2["state"] =  DISABLED
    btn1["bg"] = "white"
    btn1["fg"] = "#6B6B6B"
    btn2["bg"] = "white"
    btn2["fg"] = "#6B6B6B"
    btn3["bg"] = "#ffc000"
    btn3["fg"] = "#6B6B6B"
    btn3["text"] = "Recording Push Back...(Stop Recording)"
    logging.info("Recording PBR file: "+fpath)


## Core functions
# Simconnect link
def simconnectLink() :
    global sm
    global simconnected
    global aq
    global ae
    global tugtgl
    #Local var used to loop on SimConnect link attempt
    smconnected = 0
    logging.info("SimConnect:Linking to MSFS2020...")
    while True:
        time.sleep(10)
        try:
            sm.get_paused()
            smconnected = 1
            lbl2["fg"] = "#adf542"
            lbl2["text"] = "SimConnect: Linked"
        except:
            logging.warning("Simconnect:Connecting...")
            try:
                sm = SimConnect()
                smconnected = 1
                logging.info("SimConnect:Linked")
                lbl2["fg"] = "#adf542"
                lbl2["text"] = "SimConnect: Linked"
                aq = AircraftRequests(sm) #Aircraft Requests Variable
                ae = AircraftEvents(sm) # Aircraft Event Variables
                tugtgl = ae.find("TOGGLE_PUSHBACK")
                return (sm)
            except:
                lbl2["fg"] = "#ffc000"
                lbl2["text"] = "SimConnect: Not linked"
                logging.info("Retrying in 10sec")
                continue

# Manual Push Back
def pb():
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    keyboard.add_hotkey("left", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-27)%360*11930464)))
    keyboard.add_hotkey("left", lambda:heading('left'))
    keyboard.add_hotkey("right", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+27)%360*11930464)))
    keyboard.add_hotkey("right", lambda:heading('right'))
    keyboard.add_hotkey("up", lambda:pbst())
    keyboard.add_hotkey("down", lambda:tgl())
    keyboard.add_hotkey("down", lambda:pbkstate())
    return

# Record Push Back 
def pbrec():
    global recdata
    recdata = []
    recdata.append(str(aq.get("PLANE_LATITUDE"))+"\n")
    recdata.append(str(aq.get("PLANE_LONGITUDE"))+"\n")
    recdata.append(str(aq.get("PLANE_HEADING_DEGREES_TRUE"))+"\n")
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    pbr = open(recdata[0].replace('\n', '').replace('.', '')+recdata[1].replace('\n', '').replace('.', '')+'.pbr', 'w')
    keyboard.add_hotkey("left", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-27)%360*11930464)))
    keyboard.add_hotkey("left", lambda:heading('left'))
    keyboard.add_hotkey("right", lambda:tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+27)%360*11930464)))
    keyboard.add_hotkey("right", lambda:heading('right'))
    keyboard.add_hotkey("up", lambda:pbst())
    keyboard.add_hotkey("down", lambda:tgl())
    keyboard.add_hotkey("down", lambda:recstate())
    while rec > 0:
        if aq.get("GROUND_VELOCITY") > 0.1:
            time.sleep(rft)
            recdata.append(str(int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE"))%360*11930464))+"\n")
    pbr.writelines(recdata)
    pbr.close()
    sm.sendText("Push back recorded and saved to file")
    return

# Push Back state
def recstate():
    global rec
    rec = rec - 1
    if (rec == 1):
        sm.sendText("Recording Push back...")
    return rec

# Auto-Push Back 
def pbplay(fpath):
    tgl = ae.find("TOGGLE_PUSHBACK")
    tug = ae.find("KEY_TUG_HEADING")
    pbrd = open(fpath, 'r') 
    Lines = pbrd.readlines()
    aq.set("PLANE_LATITUDE", float(Lines[0]))
    aq.set("PLANE_LONGITUDE", float(Lines[1]))
    aq.set("PLANE_HEADING_DEGREES_TRUE", float(Lines[2]))
    time.sleep(2)
    sm.sendText("Release parking brakes")
    tgl()
    count = 0
    contact = 1
    while contact > 0:
        if aq.get("GROUND_VELOCITY") > 0.1:
            contact = 0
    for line in Lines:
        count += 1
        if count > 3:
            time.sleep(rft)
            tug(int(line.strip()))
    time.sleep(erft)
    sm.sendText("Set parking brakes")
    tgl()
    
# Push Back last steering direction 
def pbst():
    tug = ae.find("KEY_TUG_HEADING")
    if (hdg == "left"):
        tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))-3)%360*11930464))
    elif (hdg == "right"):
        tug(((int(math.degrees(aq.get("PLANE_HEADING_DEGREES_TRUE")))+3)%360*11930464))
    return

# Push Back last heading update
def heading(trn):
    global hdg
    hdg = trn
    return hdg

## TODO: change def pbkstate to def pbnotify
# Push Back SM notification (Parking Brakes)
def pbkstate():
    global pbstp
    if (pbstp == 0):
        sm.sendText("Release parking brakes")
        pbstp = 1
    else:
        sm.sendText("Set parking brakes")
        pbstp = 0



    
# GUI creation
root = Tk()

# Creates main window
root.geometry('395x390')
root.configure(background='#6B6B6B')
root.title('Push Back Recorder')
root.resizable(width=False, height=False)
root.attributes('-topmost', True)
root.update()


# GUI Buttons
global btn1
global btn2
global btn3
global btn1ab
global btn2ab
global btn3ab
# Creates Main Page buttons 
btn1 = Button(root, text='Auto-Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction)
btn1.place(x=5, y=12)
btn2 = Button(root, text='Start Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction2)
btn2.place(x=5, y=52)
btn3 = Button(root, text='Record Push Back', bg='#6B6B6B', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction3)
btn3.place(x=5, y=92)
# Declares Auto-Push Back buttons
btn1ab = Button(root, text='Left', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction2)
btn2ab = Button(root, text='Straight', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction3)
btn3ab = Button(root, text='Right', state=DISABLED, bg='white', fg="white", font=('arial', 10, 'normal'), command=btnClickFunction3)


# GUI Canvas images
global HowToImg
global HowToImgTail
# Creates "How to manual push back" canvas image 
HowToImg= Canvas(root, height=190, width=380, bd=0, highlightthickness=0, relief='ridge')
picture_file = PhotoImage(file = 'howto.gif')
HowToImg.create_image(380, 0, anchor=NE, image=picture_file)
HowToImg.place(x=5, y=152)
# Declares "How to auto-push back" canvas image
HowToImgTail= Canvas(root, height=190, width=380, bd=0, highlightthickness=0, relief='ridge')
picture_fileT = PhotoImage(file = 'howtoTail.gif')
HowToImgTail.create_image(380, 0, anchor=NE, image=picture_fileT)
# Creates "Logo" canvas image
Logo= Canvas(root, height=100, width=100, bd=0, highlightthickness=0, relief='ridge')
picture_fileL = PhotoImage(file = 'logo.gif')  
Logo.create_image(100, 0, anchor=NE, image=picture_fileL)
Logo.place(x=285, y=0)


# GUI Labels
global lbl1
global lbl2
# Creates Main Page labels
lbl1 = Label(root, text='* Remember to keep this window focused while using manual/record push back', bg='#6B6B6B', fg="#ffc000", font=('arial', 7, 'normal'))
lbl1.place(x=5, y=362)
lbl2 = Label(root, text='SimConnect: Not linked', bg='#6B6B6B', fg="#ffc000", font=('arial', 9, 'normal'))
lbl2.place(x=250, y=83)


# Declares a new thread for simconnectLink
smcheck = threading.Thread(target=simconnectLink)


# Local variables
hdg = ""
pbstp = 0
rec = 2
rft = 0.300 # Refresh rate
erft = 15   #Error Bug refresh rate


#smcheck.start()
root.mainloop()
logging.info("GUI:Exit")
sm.exit()
logging.info("SimConnect:Clean exit")
os._exit(os.EX_OK)
